pkg/platform/update-api/platform.go

```diff
@@ -0,0 +1,163 @@
+	"encoding/json"
+	"github.com/Masterminds/semver"
+	"github.com/bottlerocket-os/bottlerocket-update-operator/pkg/logging"
+	"github.com/bottlerocket-os/bottlerocket-update-operator/pkg/platform"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+)
+
+// Assert Update-API as a platform implementor.
+var _ platform.Platform = (*Platform)(nil)
+
+type Platform struct {
+	log        logging.Logger
+	httpClient http.Client
+}
+
+func New() (*Platform, error) {
+	return &Platform{log: logging.New("platform"), httpClient: http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("unix", bottlerocketApiSock)
+			},
+		},
+		// Set a 10 second timeout for all requests
+		Timeout: 10 * time.Second,
+	}}, nil
+}
+
+type statusResponse struct {
+	osVersion *semver.Version
+}
+
+func (sr *statusResponse) OK() bool {
+	// Bottlerocket OS version needs to be at least 0.4.1 to support the Update API
+	constraint, _ := semver.NewConstraint(">= 0.4.1")
+	return constraint.Check(sr.osVersion)
+}
+
+// Try to determine if the Bottlerocket version at least 0.4.1
+func (p Platform) Status() (platform.Status, error) {
+	response, err := p.httpClient.Get("http://unix" + "/os")
+	if err != nil {
+		p.log.WithError(err).Error("error when trying to GET '/os'")
+		return nil, err
+	}
+	var osInfo interface{}
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(body, osInfo)
+	if err != nil {
+		return nil, err
+	}
+	// Assert 'version_id' in the OS info json output as string
+	osVersion, err := semver.NewVersion(osInfo.(map[string]interface{})["version_id"].(string))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse 'version_id' field as semver")
+	}
+	return &statusResponse{osVersion: osVersion}, nil
+}
+
+type listAvailableResponse struct {
+	chosenUpdate *UpdateImage
+}
+
+func (lar *listAvailableResponse) Updates() []platform.Update {
+	if lar.chosenUpdate == nil {
+		return nil
+	}
+	updates := make([]platform.Update, 1)
+	updates[0] = lar.chosenUpdate
+	return updates
+}
+
+func (p Platform) ListAvailable() (platform.Available, error) {
+	p.log.Debug("fetching list of available updates")
+
+	// Refresh list of updates and check if there are any available
+	err := postAction(p.httpClient, "/actions/refresh-updates")
+	if err != nil {
+		return nil, err
+	}
+
+	updateStatus, err := getUpdateStatus(p.httpClient)
+	if updateStatus.MostRecentCommand.CmdType == refresh && updateStatus.MostRecentCommand.CmdStatus == Success {
+		// If there is no available update to update to
+		if updateStatus.ChosenUpdate == nil {
+			return &listAvailableResponse{chosenUpdate: nil}, err
+		}
+		return &listAvailableResponse{chosenUpdate: updateStatus.ChosenUpdate}, err
+	} else {
+		return &listAvailableResponse{chosenUpdate: nil}, errors.New("failed to refresh updates or update action performed out of band")
+	}
```

Comment:
This doesn't look super clear to me.  And it looks like you're missing the case where `err` is set.

```suggestion
	if err != nil {
		return &listAvailableResponse{}, err
	}
	if updateStatus.MostRecentCommand.CmdType != refresh || updateStatus.MostRecentCommand.CmdStatus != Success {
		return &listAvailableResponse{chosenUpdate: nil}, errors.New("failed to refresh updates or update action performed out of band")
	}
	return &listAvailableResponse{chosenUpdate: updateStatus.ChosenUpdate}, nil
```

Note that the last line handles both cases where `updateStatus.ChosenUpdate` has a value or is nil.